TELCODOCS-1861: Lifecycle Agent Enhancements 4.17 (#80111)
openshift/openshift-docs: modules/ztp-image-based-upgrade-concept.adoc

// Module included in the following assemblies:
// * edge_computing/image-based-upgrade/ztp-image-based-upgrade.adoc

:_mod-docs-content-type: CONCEPT
[id="ztp-image-based-upgrade-concept_{context}"]

> = Using the `ImageBasedGroupUpgrade` CR for the image-based upgrade with {ztp}

Reviewer: IBGU is not ztp or gitops specific. The emphasis here should be the `Group` concept. And we should make it clear that it's a CR on the hub.

> You can use a single resource, the `ImageBasedGroupUpgrade` CR, to define your stage transition policies and provide instructions for {cgu-operator} to reconcile the policies.

Reviewer: We no longer use policies. I would say something like "to manage imaged-based upgrade on a group of cluster through all stages".

[id="ztp-image-based-upgrade-supported-combinations_{context}"]
== Supported action combinations

Actions are the list of stage transitions that {cgu-operator} completes in the steps of an upgrade plan for the selected group of clusters.
Each `plan` entry in the `ImageBasedGroupUpgrade` CR is a separate step and a step contains one or several actions that share the same rollout strategy.
You can achieve more control over the rollout strategy for each action by separating actions into steps.

These actions can be combined differently in your upgrade plan and you can append subsequent steps later.
However, you must assess the overall result of the previous steps before appending a step to your plan.
Either the `Abort` or `Rollback` action must be the first action of a separate step that you append for a set of clusters that failed a previous step.

[IMPORTANT]
====
However, you cannot remove actions or steps from an ongoing plan.
====

The following table shows example plans for different levels of control over the rollout strategy:

.Example upgrade plans
[cols=3*, width="100%", options="header"]
|====
|All actions share same strategy
|Some actions share same strategy

> |All action share same strategy

Reviewer: I like the table. It explains rolloutStrategy well. It's just too wide. Any other way of doing the table with embedded scrollbar?

a|[source,yaml]
----
plan:
- actions: ["Prep", "Upgrade", "FinalizeUpgrade"]
    rolloutStrategy:
      maxConcurrency: 200
      timeout: 60
----
a|[source,yaml]
----
plan:
- actions: ["Prep", "Upgrade"]
    rolloutStrategy:
      maxConcurrency: 200
      timeout: 60
- actions: ["FinalizeUpgrade"]
    rolloutStrategy:
      maxConcurrency: 500
      timeout: 10
----
a|[source,yaml]
----
plan:
- actions: ["Prep"]
    rolloutStrategy:
      maxConcurrency: 200
      timeout: 60
- actions: ["Upgrade"]
    rolloutStrategy:
      maxConcurrency: 200
      timeout: 20
- actions: ["FinalizeUpgrade"]
    rolloutStrategy:
      maxConcurrency: 500
      timeout: 10
----
|====

[IMPORTANT]
====

> The clusters that fail one of the actions will skip the rest of the actions in the plan.

Reviewer: ```suggestion
The clusters that fail one of the actions will skip the rest of the actions in the same plan step.
```

====
//in the plan or in the step?

The {lcao} accepts the following actions:

`Prep`:: Move to the `Prep` stage.
`Upgrade`:: Move to the `Upgrade` stage.

> `FinalizeUpgrade`:: Move to the `Idle` stage to finalize a successful upgrade.

Reviewer: ```suggestion
`FinalizeUpgrade`:: Move to the `Idle` stage by finalizing the upgrade on selected clusters that completed the upgrade action.
```

`Rollback`:: Move to the `Rollback` stage.
`FinalizeRollback`:: Move to the `Idle` stage to finalize a rollback.

> `AbortOnFailure`:: Move to the `Idle` stage to cancel the upgrade after a failed `Prep` or `Upgrade` action.

Reviewer: ```suggestion
`AbortOnFailure`:: Move to the `Idle` stage to cancel the upgrade on selected clusters that failed `Prep` or `Upgrade` action.
```

> `Abort`:: Move to the `Idle` stage to cancel the upgrade after the previous action.

Reviewer: ```suggestion
`Abort`:: Move to the `Idle` stage to cancel the upgrade on all selected clusters.
```

> The {lcao} accepts the following actions:

Reviewer: lcao -> IBGU

Author: `ImageBasedGroupUpgrade` API?

[IMPORTANT]
====

> To avoid cancelling or rolling back an upgrade, ensure that either the `Abort` or `Rollback` action is the first step in a separate `ImageBasedGroupUpgrade` CR.

Reviewer: ```suggestion
To avoid cancelling or rolling back an upgrade on all selected clusters, ensure that either the `Abort` or `Rollback` action is the first action of a separate step added later on after assessing the overall result of the previous steps.
```

> `Rollback`:: Move to the `Rollback` stage on all selected clusters.

Reviewer: successfully upgraded clusters

> `Abort`:: Move to the `Idle` stage to cancel the upgrade on all selected clusters.

Reviewer: not upgraded clusters only

Author: WDYT?
> Cancel an ongoing upgrade only on clusters that are not upgraded yet by moving to the `Idle` stage.

Please note that I also rephrased these descriptions to put emphasis on **what** the action does instead of **how** (Cancel by moving back to Idle, start upgrade by moving to Upgrade, etc).
